### PR TITLE
Clarify backwards and forward compatibility wrt. Kotlin versions

### PR DESCRIPTION
**Pull request by eskatos**
Following discussions in #519 about the usage of Kotlin 1.2.

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
I don't know what your standards are regarding line wrapping in `md` files, but I try to keep my lines to ~ 100 characters long. I don't have any editors configured for auto-line wrapping.

File: doc/getting-started/Kotlin-Versions.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+## Kotlin versions and the Gradle Kotlin DSL 
+
+Gradle ships with Kotlin DSL support which uses embedded Kotlin compiler, plus fixed Kotlin `stdlib` and `reflect` dependencies. For example, Gradle 4.3 ships with Kotlin 1.1.51, as well as `stdlib` and `reflect` with the same version. The `kotlin` package from those modules is visible through the Gradle classpath.
+
+For both backward and forward compatibility, the [compatibility guarantees](https://kotlinlang.org/docs/reference/compatibility.html) provided by Kotlin apply.
+
+### Backward compatibility
+
+Our approach is to only do backwards-breaking Kotlin upgrades on a major Gradle release. We will always clearly document which Kotlin version we ship and announce upgrade plans before a major release.
```
Comment (JLLeitschuh) [inline doc/getting-started/Kotlin-Versions.md]: Is this guaranteed currently as the Kotlin DSL is currently in the `pre-release` state?

File: doc/getting-started/Kotlin-Versions.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+For both backward and forward compatibility, the [compatibility guarantees](https://kotlinlang.org/docs/reference/compatibility.html) provided by Kotlin apply.
+
+### Backward compatibility
+
+Our approach is to only do backwards-breaking Kotlin upgrades on a major Gradle release. We will always clearly document which Kotlin version we ship and announce upgrade plans before a major release. 
+
+Plugin authors who want to stay compatible with older Gradle versions need to limit their API usage to a subset that is compatible with these old versions. It’s not really different from any other new API in Gradle. E.g. if we introduce a new API for dependency resolution and a plugin wants to use that API, then they either need to drop support for older Gradle versions or they need to do some clever organization of their code to only execute the new code path on newer versions.
```
Comment (JLLeitschuh) [inline doc/getting-started/Kotlin-Versions.md]: How will API versioning be indicated on various methods/files? In the kotlin docs?
There's a library called [API Guardian](https://github.com/apiguardian-team/apiguardian) (created for and used by the Junit 5 team) that may work for this.

File: doc/getting-started/Kotlin-Versions.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+### Forward compatibility
+
+The biggest issue is the compatibility between the external `kotlin-gradle-plugin` version and the `kotlin-stdlib` version shipped with Gradle. More generally, between any plugin that transitively depends on `kotlin-stdlib` and its version shipped with Gradle. As long as the combination is compatible everything should work. This will become less of an issue as the language matures.
```
Comment (JLLeitschuh) [inline doc/getting-started/Kotlin-Versions.md]: Not entirely sure about this one but make sure you have a EOL at EOF.

Comment (eskatos) [reply doc/getting-started/Kotlin-Versions.md]: EOL at EOF is present.

File: doc/getting-started/Kotlin-Versions.md (hunk 2 above)
Comment (eskatos) [reply doc/getting-started/Kotlin-Versions.md]: Let's move this as a follow up. Could you please create an issue for it?

File: doc/getting-started/Kotlin-Versions.md (hunk 1 above)
Comment (eskatos) [reply doc/getting-started/Kotlin-Versions.md]: Good point, this needs to be clarified too. I'll come back to this.

File: doc/getting-started/Kotlin-Versions.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+## Kotlin versions and the Gradle Kotlin DSL 
+
+Gradle ships with Kotlin DSL support which uses embedded Kotlin compiler, plus fixed Kotlin `stdlib` and `reflect`
```
Comment (bamboo) [inline doc/getting-started/Kotlin-Versions.md]: ⭕️  I would rephrase the first sentence slightly to:

> Gradle Kotlin DSL ships with the embedded Kotlin compiler plus matching versions of the Kotlin `stdlib` and `reflect` libraries.

File: doc/getting-started/Kotlin-Versions.md (hunk 1 above)
Comment (eskatos) [reply doc/getting-started/Kotlin-Versions.md]: Done, see e70f9c3

File: doc/getting-started/Kotlin-Versions.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+The [compatibility guarantees](https://kotlinlang.org/docs/reference/compatibility.html) provided by Kotlin apply for
+both backward and forward compatibility.
+
+### Backward compatibility
+
+Our approach is to only do backwards-breaking Kotlin upgrades on a major Gradle release. We will always clearly document
+which Kotlin version we ship and announce upgrade plans before a major release. 
+
+> Until the Gradle Kotlin DSL v1.0 release our policy will be to ship with the latest stable Kotlin version available
```
Comment (bamboo) [inline doc/getting-started/Kotlin-Versions.md]: ⭕️  Maybe _Until the release of Gradle Kotlin DSL v1.0 our policy..._ ?

File: doc/getting-started/Kotlin-Versions.md (hunk 2 above)
Comment (JLLeitschuh) [reply doc/getting-started/Kotlin-Versions.md]: Sure.

Comment (JLLeitschuh) [reply doc/getting-started/Kotlin-Versions.md]: See #534

File: doc/getting-started/Kotlin-Versions.md (hunk 4 above)
Comment (eskatos) [reply doc/getting-started/Kotlin-Versions.md]: I like it better 👍 
Done

File: doc/getting-started/Kotlin-Versions.md (hunk 5 above)
Comment (eskatos) [reply doc/getting-started/Kotlin-Versions.md]: 👍 
Done
